starboard/shared/posix/socket_receive_multi_msg_internal.cc

```diff
@@ -0,0 +1,211 @@
+// Copyright 2024 The Cobalt Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "starboard/shared/posix/socket_receive_multi_msg_internal.h"
+
+#include <errno.h>
+#include <string.h>
+#include <sys/socket.h>
+
+#include "starboard/common/log.h"
+#include "starboard/common/socket.h"
+#include "starboard/extension/socket_receive_multi_msg.h"
+#include "starboard/shared/posix/socket_internal.h"
+
+namespace sbposix = starboard::shared::posix;
+
+namespace {
+
+/*
+  To abstract the implementation details, all data structures are allocated
+  in a single large buffer. That also helps improve memory access performance
+  and reduce heap fragmentation.
+
+  There is a method to query the size needed for the buffer, and another to
+  initialize a buffer. For the read function, only the socket and initialized
+  buffer are needed.
+
+  Layout of internal data structure to be allocated in the buffer:
+  struct SbSocketReceiveMultiMsgResult result;
+  unsigned int num_messages;
+  unsigned int message_size;
+  unsigned int control_size;
+  unsigned int size;
+  struct iovec iovecs[num_messages];
+  struct mmsghdr msgs[num_messages];
+  struct SbSocketReceiveMultiMsgPacket packets[num_messages]
+  char message_data[num_messages * messages_size];
+  char message_control[num_messages * control_size]
+*/
+
+struct SbSocketReceiveMultiMsgDataHeader {
+  struct SbSocketReceiveMultiMsgResult result;
+  unsigned int num_messages;
+  unsigned int message_size;
+  unsigned int control_size;
+  unsigned int size;
+};
+
+struct SbSocketReceiveMultiMsgDataReference {
+  SbSocketReceiveMultiMsgDataReference(char* data);
+  SbSocketReceiveMultiMsgDataReference(unsigned int num_messages,
+                                       unsigned int message_size,
+                                       unsigned int control_size,
+                                       char* data);
+
+  void InitializeReferences(char* data);
+
+  char* data(int message) const {
+    return message_data + message * header->message_size;
+  }
+  char* control(int message) const {
+    return message_control + message * header->control_size;
+  }
+
+  SbSocketReceiveMultiMsgDataHeader* header;
+  struct iovec* iovecs;
```

Comment:
ditto as above - maybe default-initialize ? This is c++ code not C, so should be fine

Reply:
Sure, it won't affect performance much and the bugs it avoids are sneaky.